h2o-algos/src/main/java/hex/glrm/GLRM.java

```diff
@@ -1131,13 +1189,37 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
 
           // Gradient wrt x_i is matrix product \grad L_{i,j}(x_i * Y_j, A_{i,j}) * Y_j'
           double[] weight = _lossFunc[j].mlgrad(xy, (int) a[j]);
-          double[][] ysub = _yt.getCatBlock(j);
-          for (int k = 0; k < _ncolX; k++) {
-            for (int c = 0; c < weight.length; c++)
-              grad[k] += cweight * weight[c] * ysub[k][c];
+ //         double[][] ysub = _yt.getCatBlock(j);
+          if (_yt._transposed) {
+            for (int k = 0; k < _ncolX; k++) {
+              for (int c = 0; c < weight.length; c++) {
+                //             grad[k] += cweight * weight[c] * ysub[k][c];
+                int cidx = _yt.getCatCidx(j, c);
+                double archtypevalues = 0;
+                archtypevalues = _yt._archetypes[cidx][k];
+                tgrad[k] += cweight * weight[c] * archtypevalues;
+              }
+            }
+          } else {
+            for (int k = 0; k < _ncolX; k++) {
+              for (int c = 0; c < weight.length; c++) {
+                //             grad[k] += cweight * weight[c] * ysub[k][c];
+                int cidx = _yt.getCatCidx(j, c);
+                double archtypevalues = 0;
+                archtypevalues = _yt._archetypes[k][cidx];
+                tgrad[k] += cweight * weight[c] * archtypevalues;
```

Comment:
(almost) same here, except that `_yt._archetypes[cidx]` can no longer be cached...